### ISSUE TITLE
Import with tf.loadLayerModel() for GRU architecture fails due to weight shape mismatch (but import in tf.keras works fine)

### ISSUE THREAD

**Issue by philipp-ghtest** [tensorflow/tfjs#2437]
Hi,

I trained a NLP model with TF2.0 using its Keras API. The model works perfectly fine in python:

Python code:
```
model = keras.models.Sequential([
    keras.layers.GRU(128, return_sequences=True, input_shape=[ None, max_id+1]),
    keras.layers.GRU(128, return_sequences=True),
    keras.layers.GRU(128),
    keras.layers.Flatten(),
    keras.layers.Dense(output_size, activation="softmax")
])

model.compile(loss="categorical_crossentropy", optimizer="adam", metrics=['accuracy'])
history = model.fit(train_tokens_X, train_target, validation_data=(valid_tokens_X, valid_target), batch_size=32, epochs=15)
model.save(os.path.join(data_dir, "prototype.h5"))
```

Then I saved it to .h5 and converted it with `tensorflowjs_converter` just as described here: https://www.tensorflow.org/js/tutorials/conversion/import_keras

This conversion also worked perfectly fine. I ensured this by converting the JSON model back to .h5 and re-importing it in Keras, and testing yielded the same results, so I'm pretty sure this is not an issue of the `tfjs_converter`.

Then I imported it in JS:

```
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<meta http-equiv="X-UA-Compatible" content="ie=edge">
<title>Js test</title>
</head>
<body>
<h1>JavaScript TF test</h1>

<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@1.0.0/dist/tf.min.js"></script>
<script>
	// init
	let model = null;
	(async () => {
	  	alert('Hello World!');
		tf.loadLayersModel('http://localhost:3000/prototype_web/model.json'); // Code crashes here with "Error: Input 0 is incompatible with layer flatten: expected min_ndim=3, found ndim=2."
	  	alert('Hello World2!');
		model.summary();
	})();
</script>

</body>
</html>
```

This yields the following error:

> Uncaught (in promise) Error: Input 0 is incompatible with layer flatten: expected min_ndim=3, found ndim=2.

which is quite weird. So I went on and removed the `Flatten`-layer in the `model.json` to see what happens, and then the following error arises:

```
 Uncaught (in promise) Error: Shape mismatch: [384] vs. [2,384]
     at variables.ts:135
     at t.write (variables.ts:98)
     at variables.ts:339
     at Array.forEach (<anonymous>)
     at sf (variables.ts:337)
     at e.loadWeights (container.ts:598)
     at models.ts:315
     at common.ts:14
     at Object.next (common.ts:14)
     at a (common.ts:14)
```

which explains the previous error.

Since `[2,384]` is the shape of the GRU's bias weight vector, I guess something's wrong with the GRUs weight loading procedure. But I'm not familiar enough with the code (and neither with JS) to debug this any further.

Any help with this, either workarounds or fixes, is very much appreciated! 

Thank you in advance for your help. (And thank you for creating tfjs in the first place. I really appreciate it!)

Btw: importing other models without GRU layers works fine, so I'm pretty sure this not due to a JS misconfiguration

PS: This is the model.json, fyi

´´´
{"format": "layers-model", "generatedBy": "keras v2.2.4-tf", "convertedBy": "TensorFlow.js Converter v1.3.2", "modelTopology":
 {"keras_version": "2.2.4-tf", "backend": "tensorflow", "model_config": {"class_name": "Sequential", "config": {"name": "sequential", "layers": 
 [{"class_name": "GRU", "config": {"name": "gru", "trainable": true, "batch_input_shape": [32, null, 63], "dtype": "float32", "return_sequences": true, "return_state": false, "go_backwards": false, "stateful": false, "unroll": false, "time_major": false, "units": 128, "activation": "tanh", "recurrent_activation": "sigmoid", "use_bias": true, "kernel_initializer": {"class_name": "GlorotUniform", "config": {"seed": null}},
 "recurrent_initializer": {"class_name": "Orthogonal", "config": {"gain": 1.0, "seed": null}},
 "bias_initializer": {"class_name": "Zeros", "config": {}},
 "kernel_regularizer": null, "recurrent_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "recurrent_constraint": null, "bias_constraint": null, "dropout": 0.0, "recurrent_dropout": 0.0, "implementation": 2, "reset_after": true}},
 {"class_name": "GRU", "config": {"name": "gru_1", "trainable": true, "dtype": "float32", "return_sequences": true, "return_state": false, "go_backwards": false, "stateful": false, "unroll": false, "time_major": false, "units": 128, "activation": "tanh", "recurrent_activation": "sigmoid", "use_bias": true, "kernel_initializer": {"class_name": "GlorotUniform", "config": {"seed": null}},
 "recurrent_initializer": {"class_name": "Orthogonal", "config": {"gain": 1.0, "seed": null}},
 "bias_initializer": {"class_name": "Zeros", "config": {}},
 "kernel_regularizer": null, "recurrent_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "recurrent_constraint": null, "bias_constraint": null, "dropout": 0.0, "recurrent_dropout": 0.0, "implementation": 2, "reset_after": true}},
 {"class_name": "GRU", "config": {"name": "gru_2", "trainable": true, "dtype": "float32", "return_sequences": false, "return_state": false, "go_backwards": false, "stateful": false, "unroll": false, "time_major": false, "units": 128, "activation": "tanh", "recurrent_activation": "sigmoid", "use_bias": true, "kernel_initializer": {"class_name": "GlorotUniform", "config": {"seed": null}},
 "recurrent_initializer": {"class_name": "Orthogonal", "config": {"gain": 1.0, "seed": null}},
 "bias_initializer": {"class_name": "Zeros", "config": {}},
 "kernel_regularizer": null, "recurrent_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "recurrent_constraint": null, "bias_constraint": null, "dropout": 0.0, "recurrent_dropout": 0.0, "implementation": 2, "reset_after": true}},
 {"class_name": "Flatten", "config": {"name": "flatten", "trainable": true, "dtype": "float32", "data_format": "channels_last"}},
 {"class_name": "Dense", "config": {"name": "dense", "trainable": true, "dtype": "float32", "units": 3732, "activation": "softmax", "use_bias": true, "kernel_initializer": {"class_name": "GlorotUniform", "config": {"seed": null}},
 "bias_initializer": {"class_name": "Zeros", "config": {}},
 "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}}]}},
 "training_config": {"loss": "categorical_crossentropy", "metrics": ["accuracy"], "weighted_metrics": null, "sample_weight_mode": null, "loss_weights": null, "optimizer_config": {"class_name": "Adam", "config": {"name": "Adam", "learning_rate": 0.001, "decay": 0.0, "beta_1": 0.9, "beta_2": 0.999, "epsilon": 1e-07, "amsgrad": false}}}},
 "weightsManifest": [{"paths": ["group1-shard1of1.bin"], "weights": [{"name": "dense/kernel", "shape": [128, 3732], "dtype": "float32"},
 {"name": "dense/bias", "shape": [3732], "dtype": "float32"},
 {"name": "gru/kernel", "shape": [63, 384], "dtype": "float32"},
 {"name": "gru/recurrent_kernel", "shape": [128, 384], "dtype": "float32"},
 {"name": "gru/bias", "shape": [2, 384], "dtype": "float32"},
 {"name": "gru_1/kernel", "shape": [128, 384], "dtype": "float32"},
 {"name": "gru_1/recurrent_kernel", "shape": [128, 384], "dtype": "float32"},
 {"name": "gru_1/bias", "shape": [2, 384], "dtype": "float32"},
 {"name": "gru_2/kernel", "shape": [128, 384], "dtype": "float32"},
 {"name": "gru_2/recurrent_kernel", "shape": [128, 384], "dtype": "float32"},
 {"name": "gru_2/bias", "shape": [2, 384], "dtype": "float32"}]}]}
´´´

**Comment by philipp-ghtest** [comment]
Okay, I tracked the bug down but honestly my JS knowledge is not sufficient to figure out a fix.

Something's wrong with variables.ts:335 batchSetValue(). This function is called at the end of containers.ts:564 loadModel(), after constructing the architectur from the JSON (this works fine, I checked) and loading the weights (also works fine, all shapes match).

For some reason I have trouble stepping into batchSetValue(), but I stepped into variables.ts:95 write()   which is called by batchSetValue().

Here, for all layers/weights, the shapes of this.val and newVal match EXCEPT for all the bias vectors of the GRU layers. I have no idea why, this must come from batchSetValue()


Can you reproduce this by saving a model with GRUs to .h5 (I guess you don't even need to train it, initializing the weights should be sufficient) and then loading it in TFJS with tf.loadLayersModel?

**Comment by philipp-ghtest** [comment]
Turns out there's something wrong already with the variablesAndValues array at variables.ts:335 batchSetValue().
![tfjs_debug](https://user-images.githubusercontent.com/50402778/69723424-c7abc880-1119-11ea-95d1-6fb688216778.png)

**Comment by philipp-ghtest** [comment]
wait, no, it's somewhere else. sorry for the confusion. I'm still tracking it down

**Comment by philipp-ghtest** [comment]
okay, I'm gonna close this issue und reopen a more precise one
